pkg/storage/s3_backend.go

```diff
@@ -0,0 +1,181 @@
+package storage
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"os"
+
+	"github.com/mattermost/backup-restore-tool/pkg/backuprestore"
+	s3 "github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/credentials"
+	"github.com/minio/minio-go/v7/pkg/encrypt"
+	"github.com/pkg/errors"
+)
+
+type S3FileBackend struct {
+	endpoint  string
+	accessKey string
+	secretKey string
+	secure    bool
+	region    string
+	bucket    string
+	isBifrost bool
+
+	client *s3.Client
+}
+
+func NewS3FileBackend(config backuprestore.StorageConfig) (*S3FileBackend, error) {
+	backend := &S3FileBackend{
+		endpoint:  config.Endpoint,
+		accessKey: config.AccessKey,
+		secretKey: config.SecretKey,
+		secure:    config.EnableTLS,
+		region:    config.Region,
+		bucket:    config.Bucket,
+		isBifrost: config.Bifrost,
+	}
+
+	err := backend.initClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init client")
+	}
+
+	return backend, nil
+}
+
+func (b *S3FileBackend) initClient() error {
+	var creds *credentials.Credentials
+
+	if b.isBifrost {
+		creds = credentials.New(customProvider{})
+	} else {
+		creds = credentials.NewStaticV4(b.accessKey, b.secretKey, "")
+	}
+
+	opts := s3.Options{
+		Creds:  creds,
+		Secure: b.secure,
+		Region: b.region,
+	}
+
+	// If using Bifrost, override the default transport.
+	if b.isBifrost {
```

Comment:
Out of pure curiosity, do you foresee any issues with running the backup and restore process through our Bifrost pods? I don't see any immediate issues, but am wondering if we may want to tweak something about that deployment.

Reply:
Not really. It does not cache the objects during the upload or anything like that, so I think we should be good here. Not sure how much pressure it can put on the bifrost but I think that is something we will need to observe when we have it running.